compatibility-verifier/checkoutAndBuild.sh

```diff
@@ -86,67 +100,145 @@ while [ "$1" != "" ]; do
-  # Newer commit must also be null, so take the previous commit as the older one.
-  olderCommit=`git log  --pretty=format:'%h' -n 2|tail -1`
-fi
 
+# Validate and create working directory; set up build path names
 if [ -z $workingDir ]; then
   echo "Working directory must be specified"
   usage
   exit 1
 fi
-
 if [ -d $workingDir ]; then
   echo "Directory ${workingDir} already exists. Use a new directory."
   usage
   exit 1
 fi
-
 mkdir -p ${workingDir}
 if [ $? -ne 0 ]; then
   echo "Could not create ${workingDir}"
   exit 1
 fi
 workingDir=$(absPath "$workingDir")
-
+cmdDir=$(absPath "$cmdDir")
+oldBuildOutFile="${workingDir}/oldBuild.out"
+oldTargetDir="$workingDir"/oldTargetDir
+newBuildOutFile="${workingDir}/newBuild.out"
 newTargetDir="$workingDir"/newTargetDir
+curBuildOutFile="${workingDir}/currentBuild.out"
+
+# Find value of olderCommit hash
+if [ -z "$olderCommit" -a ! -z "$newerCommit" ]; then
+   echo "If old commit hash is not given, then new commit should not be given either"
+   usage
+   exit 1
+fi
+if [ -z "$olderCommit" ]; then
+  # Newer commit must also be null, so take the previous commit as the older one.
+  olderCommit=`git log  --pretty=format:'%h' -n 2|tail -1`
+fi
+
+buildNewTarget=0
+
 if [ -z "$newerCommit" ]; then
-  echo "Compiling current tree as newer version"
-  (cd $cmdDir/.. && mvn install package -DskipTests -Pbin-dist -D jdk.version=8 ${PINOT_MAVEN_OPTS} && mvn -pl pinot-tools package -DskipTests -Djdk.version=8 ${PINOT_MAVEN_OPTS} && mvn -pl pinot-integration-tests package -DskipTests -Djdk.version=8 ${PINOT_MAVEN_OPTS})
-  if [ $? -ne 0 ]; then
-    echo Compile failed.
-    exit 1
-  fi
-  ln -s $(absPath "${cmdDir}/..") "${newTargetDir}"
+  echo "Assuming current tree as newer version"
+  ln -s "${cmdDir}/.." "${newTargetDir}"
 else
   if ! mkdir -p "$newTargetDir"; then
     echo "Failed to create target directory ${newTargetDir}"
     exit 1
   fi
-  echo "Building the new version ..."
-  checkoutAndBuild "$newerCommit" "$newTargetDir"
-  if [ $? -ne 0 ];then
-    echo Could not build new version at "${newerCommit}"
-    exit 1
-  fi
+  echo "Checking out new version at commit \"${newerCommit}\""
+  checkOut "$newerCommit" "$newTargetDir"
+  buildNewTarget=1
 fi
 
-oldTargetDir="$workingDir"/oldTargetDir
 if ! mkdir -p "$oldTargetDir"; then
   echo "Failed to create target directory ${oldTargetDir}"
   exit 1
 fi
-echo "Building the old version ... "
-checkoutAndBuild "$olderCommit" "$oldTargetDir"
-if [ $? -ne 0 ]; then
-  echo Could not build older version at "${olderCommit}"
-  exit 1
+
+echo "Checking out old version at commit \"${olderCommit}\""
+checkOut "$olderCommit" "$oldTargetDir"
+
+# Start builds in parallel.
+# We need to build current directory and old commit for sure.
+# In addition, if the newer version is not the current one,
+# we will need to build that as well.
+echo Starting build for compat checker at ${cmdDir}
+(cd ${cmdDir}/..; build ${curBuildOutFile} 1) &
```

Comment:
Why do we need to build 3 times? Could we have more comments here?

Reply:
Comments are in line 162 through 165. If they are not clear enough, please suggest alternate wording